File: python-impl/src/feign/retry/default_retry_policy.py
"""Default retry policy implementation"""
from typing import Optional
from feign.retry.retry_policy import RetryPolicy


class DefaultRetryPolicy(RetryPolicy):
    """Default retry policy that retries on network errors (RuntimeError, OSError)"""

    def __init__(self, max_retries: int = 3, retry_interval: int = 1000):
        """Initialize the retry policy

        Args:
            max_retries: Maximum number of retry attempts (default: 3)
            retry_interval: Delay between retries in milliseconds (default: 1000ms)
        """
        self._max_retries = max_retries
        self._retry_interval = retry_interval

    def max_retries(self) -> int:
        """Get the maximum number of retries

        Returns:
            Maximum retry count
        """
        return self._max_retries

    def retry_interval(self) -> int:
        """Get the retry interval in milliseconds

        Returns:
            Retry interval in milliseconds
        """
        return self._retry_interval

    def can_retry(self, retry_count: int, last_exception: Optional[Exception]) -> bool:
        """Determine if a retry should be attempted

        Args:
            retry_count: Current retry attempt count (0-based)
            last_exception: The last exception that occurred

        Returns:
            True if retry should be attempted, False otherwise
        """
        # Check if we've exceeded max retries
        if retry_count >= self._max_retries:
            return False

        # No exception means no retry
        if last_exception is None:
            return False

        # Can retry on network errors (RuntimeError, OSError)
        return isinstance(last_exception, (RuntimeError, OSError))

    def __setattr__(self, name: str, value: object) -> None:
        """Allow modification of max_retries and retry_interval via properties"""
        if name in ["_max_retries", "_retry_interval"]:
            object.__setattr__(self, name, value)

File: java-impl/src/main/java/com/feign/framework/retry/DefaultRetryPolicy.java
package com.feign.framework.retry;

import java.io.IOException;

/**
 * Default retry policy implementation.
 * Retries on RuntimeException, IOException, and FeignException.
 */
public class DefaultRetryPolicy implements RetryPolicy {
    private int maxRetries = 3;
    private long retryInterval = 1000L;
    private String loadBalancerType = "ROUND_ROBIN";
    private boolean enabled = true;

    @Override
    public boolean canRetry(Exception e, int retryCount) {
        if (!enabled) {
            return false;
        }

        if (retryCount >= maxRetries) {
            return false;
        }

        // Can retry on runtime exceptions
        if (e instanceof RuntimeException) {
            return true;
        }

        // Can retry on IOException
        if (e instanceof IOException) {
            return true;
        }

        // Can retry on FeignException
        if (e instanceof com.feign.framework.FeignException) {
            return true;
        }

        return false;
    }

    @Override
    public int getMaxRetries() {
        return maxRetries;
    }

    @Override
    public long getRetryInterval() {
        return retryInterval;
    }

    @Override
    public String getLoadBalancerType() {
        return loadBalancerType;
    }

    @Override
    public boolean isEnabled() {
        return enabled;
    }

    public void setMaxRetries(int maxRetries) {
        this.maxRetries = maxRetries;
    }

    public void setRetryInterval(long retryInterval) {
        this.retryInterval = retryInterval;
    }

    public void setLoadBalancerType(String loadBalancerType) {
        this.loadBalancerType = loadBalancerType;
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
    }
}
